cpp/roaring64map.hh

```diff
@@ -0,0 +1,796 @@
+
+        if (start_high == end_high) {
+            roarings[start_high].flip(start_low, end_low);
+            return;
+        }
+        roarings[start_high].flip(start_low, std::numeric_limits<uint32_t>::max());
+        roarings[start_high++].setCopyOnWrite(copyOnWrite);
+
+        for ( ; start_high <= highBytes(range_end)-1; ++start_high) {
+            roarings[start_high].flip(std::numeric_limits<uint32_t>::min(),
+                                      std::numeric_limits<uint32_t>::max());
+            roarings[start_high].setCopyOnWrite(copyOnWrite);
+        }
+
+        roarings[start_high].flip(std::numeric_limits<uint32_t>::min(), end_low);
+        roarings[start_high].setCopyOnWrite(copyOnWrite);
+    }
+
+    /**
+     *  Remove run-length encoding even when it is more space efficient
+     *  return whether a change was applied
+     */
+    bool removeRunCompression() {
+        return std::accumulate(roarings.begin(), roarings.end(), false,
+            [](bool previous, std::pair<const uint32_t, Roaring>& map_entry) {
+                return map_entry.second.removeRunCompression() && previous;
+            });
+    }
+
+    /** convert array and bitmap containers to run containers when it is more
+     * efficient;
+     * also convert from run containers when more space efficient.  Returns
+     * true if the result has at least one run container.
+     * Additional savings might be possible by calling shrinkToFit().
+     */
+    bool runOptimize() {
+        return std::accumulate(roarings.begin(), roarings.end(), false,
+            [](bool previous, std::pair<const uint32_t, Roaring>& map_entry) {
+                return map_entry.second.runOptimize() && previous;
+            });
+    }
+
+    /**
+     * If needed, reallocate memory to shrink the memory usage. Returns
+     * the number of bytes saved.
+    */
+    size_t shrinkToFit() {
+        size_t savedBytes = 0;
+        auto iter = roarings.begin();
+        while (iter != roarings.cend()) {
+            if (iter->second.isEmpty()) {
+                // empty Roarings are 84 bytes
+                savedBytes += 88;
+                roarings.erase(iter++);
+            } else {
+                savedBytes += iter->second.shrinkToFit();
+                iter++;
+            }
+        }
+        return savedBytes;
+    }
+
+    /**
+     * Iterate over the bitmap elements. The function iterator is called once
+     * for
+     *  all the values with ptr (can be NULL) as the second parameter of each
+     * call.
+     *
+     *  roaring_iterator64 is simply a pointer to a function that returns void,
+     *  and takes (uint64_t,void*) as inputs.
+     */
+    void iterate(roaring_iterator64 iterator, void *ptr) const {
+        std::for_each(roarings.begin(), roarings.cend(),
+            [=](const std::pair<uint32_t, Roaring>& map_entry) {
+                roaring_iterate64(map_entry.second.roaring, iterator, uint64_t(map_entry.first) << 32, ptr);
```

Comment:
I added an inline function that stitches two 32-bit values back into a 64-bit one